File: ticker/basket.go
package ticker

import (
	"encoding/json"
	"fmt"
	"io/ioutil"
	"log"
	"os"
	"path/filepath"
	"sync"
	"time"
)

// TickerBasket represents a collection of related ticker symbols
type TickerBasket struct {
	ID          string   `json:"id"`
	Name        string   `json:"name"`
	Description string   `json:"description"`
	Symbols     []string `json:"symbols"`
	CreatedAt   string   `json:"created_at"`
	UpdatedAt   string   `json:"updated_at"`
	CreatedBy   string   `json:"created_by,omitempty"`
	Tags        []string `json:"tags,omitempty"`
	IsActive    bool     `json:"is_active"`
	Category    string   `json:"category,omitempty"`
}

// BasketManager manages ticker baskets
type BasketManager struct {
	dataDir string
	baskets map[string]*TickerBasket
	mutex   sync.RWMutex
}

// NewBasketManager creates a new basket manager
func NewBasketManager(dataDir string) (*BasketManager, error) {
	// Create data directory if it doesn't exist
	if err := os.MkdirAll(dataDir, 0755); err != nil {
		return nil, fmt.Errorf("failed to create data directory: %w", err)
	}

	manager := &BasketManager{
		dataDir: dataDir,
		baskets: make(map[string]*TickerBasket),
	}

	// Load existing baskets
	if err := manager.loadBaskets(); err != nil {
		log.Printf("Warning: Failed to load baskets: %v", err)
	}

	return manager, nil
}

// loadBaskets loads all basket files from the data directory
func (m *BasketManager) loadBaskets() error {
	basketDir := filepath.Join(m.dataDir, "baskets")
	if err := os.MkdirAll(basketDir, 0755); err != nil {
		return fmt.Errorf("failed to create baskets directory: %w", err)
	}

	files, err := ioutil.ReadDir(basketDir)
	if err != nil {
		return fmt.Errorf("failed to read baskets directory: %w", err)
	}

	for _, file := range files {
		if file.IsDir() || filepath.Ext(file.Name()) != ".json" {
			continue
		}

		filePath := filepath.Join(basketDir, file.Name())
		data, err := ioutil.ReadFile(filePath)
		if err != nil {
			log.Printf("Warning: Failed to read basket file %s: %v", filePath, err)
			continue
		}

		var basket TickerBasket
		if err := json.Unmarshal(data, &basket); err != nil {
			log.Printf("Warning: Failed to parse basket file %s: %v", filePath, err)
			continue
		}

		if basket.ID == "" {
			log.Printf("Warning: Basket file %s has no ID", filePath)
			continue
		}

		m.mutex.Lock()
		m.baskets[basket.ID] = &basket
		m.mutex.Unlock()
	}

	log.Printf("Loaded %d ticker baskets from %s", len(m.baskets), basketDir)
	return nil
}

// SaveBasket saves a ticker basket to disk
func (m *BasketManager) SaveBasket(basket *TickerBasket) error {
	// Validate basket
	if basket.ID == "" {
		basket.ID = generateID()
	}
	if basket.CreatedAt == "" {
		basket.CreatedAt = time.Now().Format(time.RFC3339)
	}
	basket.UpdatedAt = time.Now().Format(time.RFC3339)

	// Add to memory cache
	m.mutex.Lock()
	m.baskets[basket.ID] = basket
	m.mutex.Unlock()

	// Save to disk
	basketDir := filepath.Join(m.dataDir, "baskets")
	if err := os.MkdirAll(basketDir, 0755); err != nil {
		return fmt.Errorf("failed to create baskets directory: %w", err)
	}

	filePath := filepath.Join(basketDir, fmt.Sprintf("%s.json", basket.ID))
	data, err := json.MarshalIndent(basket, "", "  ")
	if err != nil {
		return fmt.Errorf("failed to marshal basket: %w", err)
	}

	if err := ioutil.WriteFile(filePath, data, 0644); err != nil {
		return fmt.Errorf("failed to write basket file: %w", err)
	}

	log.Printf("Saved ticker basket %s (%s) with %d symbols", 
		basket.ID, basket.Name, len(basket.Symbols))
	return nil
}

// GetBasket retrieves a ticker basket by ID
func (m *BasketManager) GetBasket(id string) (*TickerBasket, error) {
	m.mutex.RLock()
	defer m.mutex.RUnlock()

	basket, exists := m.baskets[id]
	if !exists {
		return nil, fmt.Errorf("basket not found: %s", id)
	}

	// Return a copy to avoid race conditions
	basketCopy := *basket
	return &basketCopy, nil
}

// DeleteBasket deletes a ticker basket
func (m *BasketManager) DeleteBasket(id string) error {
	// Check if basket exists
	m.mutex.RLock()
	_, exists := m.baskets[id]
	m.mutex.RUnlock()

	if !exists {
		return fmt.Errorf("basket not found: %s", id)
	}

	// Delete from memory cache
	m.mutex.Lock()
	delete(m.baskets, id)
	m.mutex.Unlock()

	// Delete from disk
	basketDir := filepath.Join(m.dataDir, "baskets")
	filePath := filepath.Join(basketDir, fmt.Sprintf("%s.json", id))
	if err := os.Remove(filePath); err != nil {
		return fmt.Errorf("failed to delete basket file: %w", err)
	}

	log.Printf("Deleted ticker basket %s", id)
	return nil
}

// ListBaskets returns a list of all ticker baskets
func (m *BasketManager) ListBaskets() []TickerBasket {
	m.mutex.RLock()
	defer m.mutex.RUnlock()

	baskets := make([]TickerBasket, 0, len(m.baskets))
	for _, basket := range m.baskets {
		baskets = append(baskets, *basket)
	}

	return baskets
}

// AddSymbolToBasket adds a symbol to a ticker basket
func (m *BasketManager) AddSymbolToBasket(basketID, symbol string) error {
	m.mutex.Lock()
	defer m.mutex.Unlock()

	basket, exists := m.baskets[basketID]
	if !exists {
		return fmt.Errorf("basket not found: %s", basketID)
	}

	// Check if symbol already exists
	for _, s := range basket.Symbols {
		if s == symbol {
			// Symbol already exists, no need to add
			return nil
		}
	}

	// Add symbol
	basket.Symbols = append(basket.Symbols, symbol)
	basket.UpdatedAt = time.Now().Format(time.RFC3339)

	// Save to disk
	basketDir := filepath.Join(m.dataDir, "baskets")
	filePath := filepath.Join(basketDir, fmt.Sprintf("%s.json", basketID))
	data, err := json.MarshalIndent(basket, "", "  ")
	if err != nil {
		return fmt.Errorf("failed to marshal basket: %w", err)
	}

	if err := ioutil.WriteFile(filePath, data, 0644); err != nil {
		return fmt.Errorf("failed to write basket file: %w", err)
	}

	log.Printf("Added symbol %s to basket %s (%s)", symbol, basketID, basket.Name)
	return nil
}

// RemoveSymbolFromBasket removes a symbol from a ticker basket
func (m *BasketManager) RemoveSymbolFromBasket(basketID, symbol string) error {
	m.mutex.Lock()
	defer m.mutex.Unlock()

	basket, exists := m.baskets[basketID]
	if !exists {
		return fmt.Errorf("basket not found: %s", basketID)
	}

	// Find and remove symbol
	for i, s := range basket.Symbols {
		if s == symbol {
			// Remove symbol
			basket.Symbols = append(basket.Symbols[:i], basket.Symbols[i+1:]...)
			basket.UpdatedAt = time.Now().Format(time.RFC3339)

			// Save to disk
			basketDir := filepath.Join(m.dataDir, "baskets")
			filePath := filepath.Join(basketDir, fmt.Sprintf("%s.json", basketID))
			data, err := json.MarshalIndent(basket, "", "  ")
			if err != nil {
				return fmt.Errorf("failed to marshal basket: %w", err)
			}

			if err := ioutil.WriteFile(filePath, data, 0644); err != nil {
				return fmt.Errorf("failed to write basket file: %w", err)
			}

			log.Printf("Removed symbol %s from basket %s (%s)", symbol, basketID, basket.Name)
			return nil
		}
	}

	// Symbol not found
	return fmt.Errorf("symbol not found in basket: %s", symbol)
}

// GetBasketsBySymbol returns a list of baskets containing a specific symbol
func (m *BasketManager) GetBasketsBySymbol(symbol string) []TickerBasket {
	m.mutex.RLock()
	defer m.mutex.RUnlock()

	var baskets []TickerBasket
	for _, basket := range m.baskets {
		for _, s := range basket.Symbols {
			if s == symbol {
				baskets = append(baskets, *basket)
				break
			}
		}
	}

	return baskets
}

// GetSymbolsFromBaskets returns a unique list of all symbols from a set of basket IDs
func (m *BasketManager) GetSymbolsFromBaskets(basketIDs []string) []string {
	m.mutex.RLock()
	defer m.mutex.RUnlock()

	// Use a map for deduplication
	symbolMap := make(map[string]bool)

	for _, basketID := range basketIDs {
		basket, exists := m.baskets[basketID]
		if !exists {
			continue
		}

		for _, symbol := range basket.Symbols {
			symbolMap[symbol] = true
		}
	}

	// Convert map keys to slice
	symbols := make([]string, 0, len(symbolMap))
	for symbol := range symbolMap {
		symbols = append(symbols, symbol)
	}

	return symbols
}

// Helper function to generate a simple ID
func generateID() string {
	return fmt.Sprintf("basket_%d", time.Now().UnixNano())
}

File: rust-trader/crates/ticker/src/basket.rs
//! Basket management — persistent JSON-file storage for symbol groups.

use std::collections::HashMap;
use std::path::{Path, PathBuf};
use std::sync::{Arc, RwLock};

use chrono::Utc;
use serde::{Deserialize, Serialize};
use tokio::fs;

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TickerBasket {
    pub id: String,
    pub name: String,
    pub description: String,
    pub symbols: Vec<String>,
    pub created_at: String,
    pub updated_at: String,
    #[serde(default)]
    pub created_by: Option<String>,
    #[serde(default)]
    pub tags: Vec<String>,
    pub is_active: bool,
    #[serde(default)]
    pub category: Option<String>,
}

pub struct BasketManager {
    data_dir: PathBuf,
    baskets: Arc<RwLock<HashMap<String, TickerBasket>>>,
}

impl BasketManager {
    pub async fn new(
        data_dir: impl AsRef<Path>,
    ) -> Result<Self, Box<dyn std::error::Error + Send + Sync>> {
        let data_dir = data_dir.as_ref().to_path_buf();
        fs::create_dir_all(data_dir.join("baskets")).await?;
        let mut mgr = Self {
            data_dir,
            baskets: Arc::new(RwLock::new(HashMap::new())),
        };
        mgr.load_baskets().await?;
        Ok(mgr)
    }

    async fn load_baskets(&mut self) -> Result<(), Box<dyn std::error::Error + Send + Sync>> {
        let basket_dir = self.data_dir.join("baskets");
        let _ = fs::create_dir_all(&basket_dir).await;
        let mut entries = fs::read_dir(&basket_dir).await?;
        while let Some(entry) = entries.next_entry().await? {
            let path = entry.path();
            if path.extension().map(|e| e == "json").unwrap_or(false) {
                let data = fs::read_to_string(&path).await?;
                if let Ok(basket) = serde_json::from_str::<TickerBasket>(&data) {
                    if !basket.id.is_empty() {
                        self.baskets
                            .write()
                            .unwrap()
                            .insert(basket.id.clone(), basket);
                    }
                }
            }
        }
        Ok(())
    }

    pub async fn save_basket(
        &self,
        mut basket: TickerBasket,
    ) -> Result<(), Box<dyn std::error::Error + Send + Sync>> {
        if basket.id.is_empty() {
            basket.id = format!("basket_{}", Utc::now().timestamp_nanos_opt().unwrap_or(0));
        }
        if basket.created_at.is_empty() {
            basket.created_at = Utc::now().to_rfc3339();
        }
        basket.updated_at = Utc::now().to_rfc3339();
        let dir = self.data_dir.join("baskets");
        fs::create_dir_all(&dir).await?;
        let path = dir.join(format!("{}.json", basket.id));
        fs::write(path, serde_json::to_string_pretty(&basket)?).await?;
        self.baskets
            .write()
            .unwrap()
            .insert(basket.id.clone(), basket);
        Ok(())
    }

    pub fn get_basket(&self, id: &str) -> Option<TickerBasket> {
        self.baskets.read().unwrap().get(id).cloned()
    }

    pub async fn delete_basket(
        &self,
        id: &str,
    ) -> Result<bool, Box<dyn std::error::Error + Send + Sync>> {
        let existed = self.baskets.write().unwrap().remove(id).is_some();
        if existed {
            let _ =
                fs::remove_file(self.data_dir.join("baskets").join(format!("{}.json", id))).await;
        }
        Ok(existed)
    }

    pub fn list_baskets(&self) -> Vec<TickerBasket> {
        self.baskets.read().unwrap().values().cloned().collect()
    }

    pub fn get_symbols_from_baskets(&self, ids: &[String]) -> Vec<String> {
        let guard = self.baskets.read().unwrap();
        let mut set = std::collections::HashSet::new();
        for id in ids {
            if let Some(b) = guard.get(id) {
                for s in &b.symbols {
                    set.insert(s.clone());
                }
            }
        }
        set.into_iter().collect()
    }
}
